core/tests/protractor_desktop/additionalEditorFeaturesModals.js

```diff
@@ -380,6 +380,148 @@ describe('Full exploration editor', function() {
       await users.logout();
     });
 
+  it(
+    'should show a toast message to inform when the user is offline or online',
+    async function() {
+      await users.createUser('user16@editor.com', 'user16Editor');
+
+      await users.login('user16@editor.com');
+      await workflow.createExploration(true);
+      await explorationEditorPage.navigateToMainTab();
+
+      // Set network connection to offline.
+      await browser.driver.setNetworkConditions(
+        {
+          offline: true,
+          latency: 1000,
+          download_throughput: 0,
+          upload_throughput: 0});
+
+      // Check that toast message appeared when offline.
+      await waitFor.visibilityOf(
+        element(by.css('.protractor-test-toast-message')),
+        'Offline warning toast message taking too long to appear.');
+
+      // Set network connection to online.
+      await browser.driver.setNetworkConditions(
+        {
+          offline: false,
+          latency: 150,
+          download_throughput: 450 * 1024,
+          upload_throughput: 150 * 1024});
+
+      // Check that toast message appeared when reconnected.
+      await waitFor.visibilityOf(
+        element(by.css('.protractor-test-toast-message')),
+        'Online warning toast message taking too long to appear.');
+      await users.logout();
+    });
+
+  it(
+    'should be able to save the changes when the user is offline',
+    async function() {
+      await users.createUser('user17@editor.com', 'user17Editor');
+
+      await users.login('user17@editor.com');
+      await workflow.createExploration(true);
+      await explorationEditorPage.navigateToMainTab();
+
+      // Set network connection to offline.
+      await browser.driver.setNetworkConditions(
+        {
+          offline: true,
+          latency: 1000,
+          download_throughput: 0,
+          upload_throughput: 0});
+
+      await waitFor.visibilityOf(
+        element(by.css('.protractor-test-toast-message')),
+        'Offline warning toast message taking too long to appear.');
+
+      // Add a content change to check changes can be done when offline.
+      await explorationEditorMainTab.setContent(async function(richTextEditor) {
+        await richTextEditor.appendPlainText('How are you feeling?');
+      });
+
+      // Set network connection to online.
+      await browser.driver.setNetworkConditions(
+        {
+          offline: false,
+          latency: 150,
+          download_throughput: 450 * 1024,
+          upload_throughput: 150 * 1024});
+      await waitFor.visibilityOf(
+        element(by.css('.protractor-test-toast-message')),
+        'Online warning toast message taking too long to appear.');
+
+      // Check that the changes are auto-saved when reconnected.
+      await action.waitForAutosave();
```

Comment:
This doesn't actually check that the changes are autosaved since `wiatForAutosave` just waits for the autosave indicator to disappear. This will be a no-op if the autosave indicator isn't present in the first place

https://github.com/oppia/oppia/blob/e41fb8a901eff98e44546d300275e549b150377b/core/tests/protractor_utils/action.js#L25-L28

Reply:
Yeah I get it. But that autosave will always appear so we are just checking if it has disappeared to move forward and also at last we are checking that whether the changes made offline are applied or not.